src/v/kafka/requests/fetch_request.cc

```diff
@@ -331,79 +335,92 @@ read_from_ntp(op_context& octx, model::ntp ntp, fetch_config config) {
      * the tp in the metadata cache so that this condition is unlikely
      * to pass.
      */
-    const auto mntpv = model::materialized_ntp(std::move(ntp));
+    auto mntpv = model::materialized_ntp(std::move(ntp));
     auto shard = octx.rctx.shards().shard_for(mntpv.source_ntp());
 
     if (unlikely(!shard)) {
         return make_ready_partition_response_error(
           error_code::unknown_topic_or_partition);
     }
-
-    return octx.rctx.partition_manager().invoke_on(
-      *shard,
-      octx.ssg,
-      [initial_fetch = octx.initial_fetch,
-       deadline = octx.deadline,
-       mntpv = std::move(mntpv),
-       config](cluster::partition_manager& mgr) {
-          /*
-           * lookup the ntp's partition
-           */
-          auto partition = mgr.get(mntpv.source_ntp());
-          if (unlikely(!partition)) {
-              return make_ready_partition_response_error(
-                error_code::unknown_topic_or_partition);
-          }
-          if (unlikely(!partition->is_leader())) {
-              return make_ready_partition_response_error(
-                error_code::not_leader_for_partition);
-          }
-          if (mntpv.is_materialized()) {
-              if (auto log = mgr.log(mntpv.input_ntp())) {
-                  return read_from_partition(
-                    partition_wrapper(partition, log), config, std::nullopt);
-              } else {
-                  return make_ready_partition_response_error(
-                    error_code::unknown_topic_or_partition);
-              }
-          }
-
-          auto high_watermark = partition->high_watermark();
-          auto max_offset = high_watermark < model::offset(0)
-                              ? model::offset(0)
-                              : high_watermark + model::offset(1);
-          if (
-            config.start_offset < partition->start_offset()
-            || config.start_offset > max_offset) {
-              return ss::make_ready_future<fetch_response::partition_response>(
-                fetch_response::partition_response{
-                  .error = error_code::offset_out_of_range,
-                  .high_watermark = model::offset(-1),
-                  .last_stable_offset = model::offset(-1),
-                  .log_start_offset = model::offset(-1),
-                  .record_set = iobuf(),
-                });
+    bool foreign_read = shard != ss::this_shard_id();
```

Comment:
this seems like a good candidate for splitting change and formatting into separate patches.

Reply:
I know, but this patch itself doesn't add any value into the fetch request handing. there are other issues that would make it incorrect. this is why I decided to make it part of this series.

Reply:
done